Mathlib/RingTheory/NoetherNormalization.lean

```diff
@@ -0,0 +1,401 @@
+    obtain ⟨pi, pj⟩ := Fin.find_eq_some_iff.mp hi
+    unfold p at pj
+    set a0 := ((v 0 : ℤ) - w 0)
+    set a :=
+      fun (j : Fin n) ↦ (((n + 1 : ℤ) * up f) ^ (n + 1 - j.succ) * ((v j.succ) - (w j.succ)))
+    have eq2 : - a i = a0 + ∑ j : Fin n with j ≠ i, a j := by
+      rw [← add_zero (- a i), ← equal f v w h, ← add_assoc]
+      nth_rw 2 [add_comm]
+      rw [add_assoc, add_left_cancel_iff]
+      have : a i = ∑ j : Fin n with j = i, a j := by
+        simp only [Finset.sum_filter, Finset.sum_ite_eq', Finset.mem_univ, if_pos]
+      simp only [this, ← Finset.sum_neg_distrib, Finset.sum_filter, ← Finset.sum_add_distrib]
+      apply Finset.sum_congr rfl
+      intro j _
+      by_cases hij : j = i
+      · simp only [if_pos hij, if_neg (fun (t : j ≠ i) ↦ (t hij)), neg_add_cancel]
+      · rw [if_pos hij, if_neg hij, neg_zero, zero_add]
+    have : |- a i| > |a0| + ∑ j : Fin n with j ≠ i, |a j| := by
+      rw [Finset.sum_filter]
+      set bi := (n + 1 : ℤ) ^ (n - i.succ) * up f ^ (n + 1 - i.succ)
+      have : |a0| < up f := by
+        unfold a0
+        exact abs_sub_lt_of_nonneg_of_lt (Int.ofNat_zero_le _) (cast_lt.mpr (vlt 0))
+          (Int.ofNat_zero_le _) (cast_lt.mpr (wlt 0))
+      have ai : |- a i| ≥ (n + 1) * bi * 1 := by
+        unfold a bi
+        simp only [abs_neg, abs_mul, mul_pow]
+        have : |v i.succ - (w i.succ : ℤ)| ≥ 1 := by apply Int.one_le_abs; omega
+        nth_rw 1 [Nat.sub_add_comm (Fin.is_le _), add_comm (n - i.succ), pow_add, pow_one]
+        have n1 : (n : ℤ) + 1 = (@Nat.cast ℤ) (n + 1) := rfl
+        simp only [abs_mul, abs_pow, n1, abs_cast, ← mul_assoc]
+        apply mul_le_mul (by rfl) (by linarith) Int.one_nonneg
+        apply Int.mul_nonneg (Int.mul_nonneg (Int.ofNat_zero_le _ ) (pow_nonneg
+          (Int.ofNat_zero_le _) (n - i.succ))) (pow_nonneg (Int.ofNat_zero_le _) (n + 1- ↑i.succ))
+      have aj := a_j _ _ _ vlt wlt i pj
+      rw [mul_one, add_mul, one_mul, add_comm] at ai
+      have : ∑ j : Fin n, (if j ≠ i then |a j| else 0) ≤ ∑ _ : Fin n, bi :=
+        Finset.sum_le_sum (fun i _ ↦ aj i)
+      rw [Fin.sum_const, nsmul_eq_mul] at this
+      have : bi ≥ up f := by
+        unfold bi
+        simp only [succ_eq_add_one, Fin.val_succ, reduceSubDiff, ge_iff_le]
+        nth_rw 1 [← one_mul (up f : ℤ)]
+        have le := one_le_pow₀ (le_add_of_nonneg_left (Int.ofNat_zero_le n)) (n := n - (i + 1))
+        apply Int.mul_le_mul le ?_ (Int.ofNat_zero_le _) (by linarith)
+        apply le_self_pow₀ ?_ (sub_ne_zero_of_lt i.isLt)
+        simp only [cast_add, cast_one, le_add_iff_nonneg_right]
+        exact Int.ofNat_zero_le _
+      linarith
+    rw [eq2] at this
+    have h2 := abs_add_le a0 (∑ j : Fin n with j ≠ i, a j)
+    have h3 := Finset.abs_sum_le_sum_abs a (Finset.filter (fun j ↦ j ≠ i) Finset.univ)
+    linarith
+
+private lemma T_spec_monomial (a : k) (ha : a ≠ 0) : ((T f) (monomial v a)).degreeOf 0 =
+    ∑ i : Fin (n + 1), (r f i) * ((monomial v a).degreeOf i) := by
```

Comment:
`(monomial v a).degreeOf i` is the same as `v i` by [MvPolynomial.degreeOf_monomial_eq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/MvPolynomial/Degrees.html#MvPolynomial.degreeOf_monomial_eq). I think you should use the shorter spelling in the statement.